File: blueprints/search/tools/x-viewer/src/tid.ts
const tidKeyword = 'obfiowerehiring'
const tidPairsURL = 'https://raw.githubusercontent.com/fa0311/x-client-transaction-id-pair-dict/refs/heads/main/pair.json'
const tidEpochOffset = 1682924400

interface TIDPair {
  animationKey: string
  verification: string
}

let cachedPairs: TIDPair[] = []
let lastFetch = 0

async function fetchTIDPairs(): Promise<TIDPair[]> {
  const now = Date.now()
  if (cachedPairs.length > 0 && now - lastFetch < 3600000) {
    return cachedPairs
  }

  try {
    const resp = await fetch(tidPairsURL)
    if (!resp.ok) {
      if (cachedPairs.length > 0) return cachedPairs
      throw new Error(`fetch TID pairs: ${resp.status}`)
    }
    const pairs: TIDPair[] = await resp.json()
    if (pairs.length === 0) {
      if (cachedPairs.length > 0) return cachedPairs
      throw new Error('TID pairs empty')
    }
    cachedPairs = pairs
    lastFetch = now
    return pairs
  } catch (e) {
    if (cachedPairs.length > 0) return cachedPairs
    throw e
  }
}

function base64Decode(s: string): Uint8Array {
  let padded = s
  const m = padded.length % 4
  if (m !== 0) padded += '='.repeat(4 - m)
  const binary = atob(padded)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i)
  }
  return bytes
}

function base64Encode(bytes: Uint8Array): string {
  let binary = ''
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i])
  }
  return btoa(binary).replace(/=+$/, '')
}

export async function generateTID(path: string): Promise<string> {
  const pairs = await fetchTIDPairs()
  const pair = pairs[Math.floor(Math.random() * pairs.length)]

  const timeNow = Math.floor(Date.now() / 1000) - tidEpochOffset
  const timeNowBytes = new Uint8Array([
    timeNow & 0xff,
    (timeNow >> 8) & 0xff,
    (timeNow >> 16) & 0xff,
    (timeNow >> 24) & 0xff,
  ])

  const data = `GET!${path}!${timeNow}${tidKeyword}${pair.animationKey}`
  const encoder = new TextEncoder()
  const hashBuffer = await crypto.subtle.digest('SHA-256', encoder.encode(data))
  const hashBytes = new Uint8Array(hashBuffer)

  const keyBytes = base64Decode(pair.verification)

  const bytesArr = new Uint8Array(keyBytes.length + 4 + 16 + 1)
  bytesArr.set(keyBytes, 0)
  bytesArr.set(timeNowBytes, keyBytes.length)
  bytesArr.set(hashBytes.slice(0, 16), keyBytes.length + 4)
  bytesArr[keyBytes.length + 4 + 16] = 3

  const randomNum = Math.floor(Math.random() * 256)
  const tid = new Uint8Array(1 + bytesArr.length)
  tid[0] = randomNum
  for (let i = 0; i < bytesArr.length; i++) {
    tid[i + 1] = bytesArr[i] ^ randomNum
  }

  return base64Encode(tid)
}

File: blueprints/search/pkg/scrape/x/tid.go
package x

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/rand/v2"
	"net/http"
	"strings"
	"sync"
	"time"
)

// TID generation ported from Nitter's src/tid.nim.
// Generates x-client-transaction-id headers required for API access.

const (
	tidKeyword = "obfiowerehiring"
	tidPairsURL = "https://raw.githubusercontent.com/fa0311/x-client-transaction-id-pair-dict/refs/heads/main/pair.json"
	tidCacheTTL = 1 * time.Hour
	// Epoch offset: April 30, 2023 21:00:00 UTC
	tidEpochOffset = 1682924400
)

type tidPair struct {
	AnimationKey string `json:"animationKey"`
	Verification string `json:"verification"`
}

var (
	tidMu         sync.Mutex
	tidPairs      []tidPair
	tidLastFetch  time.Time
	tidHTTPClient = &http.Client{Timeout: 15 * time.Second}
)

func fetchTIDPairs() ([]tidPair, error) {
	tidMu.Lock()
	defer tidMu.Unlock()

	if len(tidPairs) > 0 && time.Since(tidLastFetch) < tidCacheTTL {
		return tidPairs, nil
	}

	resp, err := tidHTTPClient.Get(tidPairsURL)
	if err != nil {
		if len(tidPairs) > 0 {
			return tidPairs, nil // use stale cache
		}
		return nil, fmt.Errorf("fetch TID pairs: %w", err)
	}
	defer resp.Body.Close()

	var pairs []tidPair
	if err := json.NewDecoder(resp.Body).Decode(&pairs); err != nil {
		if len(tidPairs) > 0 {
			return tidPairs, nil
		}
		return nil, fmt.Errorf("parse TID pairs: %w", err)
	}

	if len(pairs) == 0 {
		if len(tidPairs) > 0 {
			return tidPairs, nil
		}
		return nil, fmt.Errorf("TID pairs empty")
	}

	tidPairs = pairs
	tidLastFetch = time.Now()
	return pairs, nil
}

// generateTID generates an x-client-transaction-id for the given API path.
func generateTID(path string) (string, error) {
	pairs, err := fetchTIDPairs()
	if err != nil {
		return "", err
	}

	pair := pairs[rand.IntN(len(pairs))]

	timeNow := int(time.Now().Unix()) - tidEpochOffset
	timeNowBytes := []byte{
		byte(timeNow & 0xff),
		byte((timeNow >> 8) & 0xff),
		byte((timeNow >> 16) & 0xff),
		byte((timeNow >> 24) & 0xff),
	}

	data := fmt.Sprintf("GET!%s!%d%s%s", path, timeNow, tidKeyword, pair.AnimationKey)
	hashBytes := sha256.Sum256([]byte(data))

	keyBytes, err := base64.StdEncoding.DecodeString(pair.Verification)
	if err != nil {
		// Try with padding added
		padded := pair.Verification
		if m := len(padded) % 4; m != 0 {
			padded += strings.Repeat("=", 4-m)
		}
		keyBytes, err = base64.StdEncoding.DecodeString(padded)
		if err != nil {
			return "", fmt.Errorf("decode verification key: %w", err)
		}
	}

	// Build: keyBytes + timeNowBytes + hash[0:16] + [3]
	bytesArr := make([]byte, 0, len(keyBytes)+4+16+1)
	bytesArr = append(bytesArr, keyBytes...)
	bytesArr = append(bytesArr, timeNowBytes...)
	bytesArr = append(bytesArr, hashBytes[:16]...)
	bytesArr = append(bytesArr, 3)

	// XOR with random byte
	randomNum := byte(rand.IntN(256))
	tid := make([]byte, 1+len(bytesArr))
	tid[0] = randomNum
	for i, b := range bytesArr {
		tid[i+1] = b ^ randomNum
	}

	// Base64 without padding
	return strings.TrimRight(base64.StdEncoding.EncodeToString(tid), "="), nil
}
